motor/core.py

```diff
@@ -56,6 +56,9 @@
     ssl = None
     HAS_SSL = False
 
+# Surface deprecation warnings to users by default
+warnings.filterwarnings("default", category=DeprecationWarning)
```

Comment:
What are the implications of this? IIUC we are changing this option for the entire Python process, not just Motor. For example what happens when the user runs with `python -Werror ..`? Does this override their "error" setting?

What if we manually add deprecation warnings to motor itself before attempting the operation on the background thread?

Reply:
> What are the implications of this? IIUC we are changing this option for the entire Python process, not just Motor. For example what happens when the user runs with `python -Werror ..`? Does this override their "error" setting?
> 
DeprecationWarnings will still show as warnings in that case I think so it does do overriding
> What if we manually add deprecation warnings to motor itself before attempting the operation on the background thread?
What exactly does this entail?

Reply:
Motor essentially forwards method calls to a background thread which executes them via Pymongo on a background thread. You can read more about Motor's design here: https://motor.readthedocs.io/en/stable/developer-guide.html

I'm suggesting that _before_ sending a deprecated method to a background thread, we explicitly call `warnings.warn('...', DeprecationWarning, stacklevel=...)` in motor. This isn't ideal since we need to maintain these duplicate warnings in motor but it might be the best we can do. 

An alternative might be to somehow forward the warning from the background thread but I'm not sure it that is feasible.

Reply:
@prashantmital can you chime in. Is this required for 2.4? Maybe we should punt this issue to the next release?

Reply:
ok I'll leave this open for now and work on something else

Reply:
I think `sys.warnoptions` is what we need here. 

```
if not sys.warnoptions:
    warnings.simplefilter("default")
```

(see https://docs.python.org/3/library/warnings.html?highlight=warnings#overriding-the-default-filter for another example of using `sys.warnoptions`)


I am also not sure why we are restricting this to `DeprecationWarning`? Seems like this would be an issue that plagues any warnings raised by a background thread. Hence, I suggested the above solution that would do this for _all_ warnings but maybe I am missing something here.

Reply:
Prashant's fix seems to fix the cases where users specify a flag (such as `-Werror`) and allow them to display as the user wants them to

Reply:
It seems odd to me that Motor, a python library (not an application), would change the default warning filter for the entire application. [MOTOR-607](https://jira.mongodb.org/browse/MOTOR-607) says 
> Since these synchronous calls are run using a thread pool executor, the application must be run using -Wdefault in order to make the warning user-visible.

Is this behavior explained anywhere? I can't fin anything about threads not showing warnings as expected.

Reply:
We definitely cannot call `warnings.simplefilter("default")`. Wouldn't it break applications that do this:
```
warnings.simplefilter("error")
import motor
# Calling deprecated method should error but does not...
```

Reply:
I believe `DeprecationWarnings` are ignored by Python default and not anything motor is doing 
![image](https://user-images.githubusercontent.com/30269279/113613172-ab6d4500-9605-11eb-9a95-2b2371b4bd92.png)

Reply:
I'll run a quick test on this

> We definitely cannot call `warnings.simplefilter("default")`. Wouldn't it break applications that do this?:
> 
> ```
> warnings.simplefilter("error")
> import motor
> # Calling deprecated method should error but does not...
> ```

Reply:
@ShaneHarvey you're right, it does break those applications

Reply:
> Is this behavior explained anywhere? I can't fin anything about threads not showing warnings as expected.

I think @guanlinzhou is right - `DeprecationWarning`s seem to be suppressed unless they are raised in `__main__`. So it appears that this is expected behavior.